nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -64,40 +105,118 @@ std::shared_ptr<Costmap2D> CostmapSubscriber::getCostmap()
 
 void CostmapSubscriber::toCostmap2D()
 {
-  if (costmap_ == nullptr) {
-    costmap_ = std::make_shared<Costmap2D>(
-      costmap_msg_->metadata.size_x, costmap_msg_->metadata.size_y,
-      costmap_msg_->metadata.resolution, costmap_msg_->metadata.origin.position.x,
-      costmap_msg_->metadata.origin.position.y);
-  } else if (costmap_->getSizeInCellsX() != costmap_msg_->metadata.size_x ||  // NOLINT
-    costmap_->getSizeInCellsY() != costmap_msg_->metadata.size_y ||
-    costmap_->getResolution() != costmap_msg_->metadata.resolution ||
-    costmap_->getOriginX() != costmap_msg_->metadata.origin.position.x ||
-    costmap_->getOriginY() != costmap_msg_->metadata.origin.position.y)
-  {
-    // Update the size of the costmap
-    costmap_->resizeMap(costmap_msg_->metadata.size_x, costmap_msg_->metadata.size_y,
-      costmap_msg_->metadata.resolution,
-      costmap_msg_->metadata.origin.position.x,
-      costmap_msg_->metadata.origin.position.y);
-  }
+  if (use_raw_) {
+    if (costmap_ == nullptr) {
+      costmap_ = std::make_shared<Costmap2D>(
+        costmap_raw_msg_->metadata.size_x, costmap_raw_msg_->metadata.size_y,
+        costmap_raw_msg_->metadata.resolution, costmap_raw_msg_->metadata.origin.position.x,
+        costmap_raw_msg_->metadata.origin.position.y);
+    } else if (costmap_->getSizeInCellsX() != costmap_raw_msg_->metadata.size_x ||  // NOLINT
+      costmap_->getSizeInCellsY() != costmap_raw_msg_->metadata.size_y ||
+      costmap_->getResolution() != costmap_raw_msg_->metadata.resolution ||
+      costmap_->getOriginX() != costmap_raw_msg_->metadata.origin.position.x ||
+      costmap_->getOriginY() != costmap_raw_msg_->metadata.origin.position.y)
+    {
+      // Update the size of the costmap
+      costmap_->resizeMap(costmap_raw_msg_->metadata.size_x, costmap_raw_msg_->metadata.size_y,
+        costmap_raw_msg_->metadata.resolution,
+        costmap_raw_msg_->metadata.origin.position.x,
+        costmap_raw_msg_->metadata.origin.position.y);
+    }
+
+    unsigned char * master_array = costmap_->getCharMap();
+    unsigned int index = 0;
+    for (unsigned int i = 0; i < costmap_raw_msg_->metadata.size_x; ++i) {
+      for (unsigned int j = 0; j < costmap_raw_msg_->metadata.size_y; ++j) {
+        master_array[index] = costmap_raw_msg_->data[index];
+        ++index;
+      }
+    }
+  } else {
+    if (has_update_) {
+      unsigned char * master_array = costmap_->getCharMap();
+      unsigned int di = 0;
+      for (unsigned int y = 0; y < costmap_update_msg_->height; y++) {
+        unsigned int index_base = (costmap_update_msg_->y + y) * costmap_msg_->info.width;
+        for (unsigned int x = 0; x < costmap_update_msg_->width; x++) {
+          unsigned int index = index_base + x + costmap_update_msg_->x;
+          master_array[index] = interpretValue(costmap_update_msg_->data[di++]);
+        }
+      }
+    } else {
+      if (costmap_ == nullptr) {
```

Comment:
You should pull out the `if no costmap, make shared ptr` logic outside of the `if/else` for `use_raw_`. I think the only reason we need this is for some reason we duplicated the metadata https://github.com/ros-planning/navigation2/blob/master/nav2_msgs/msg/CostmapMetaData.msg when the nav_msgs version https://github.com/ros/common_msgs/blob/jade-devel/nav_msgs/msg/MapMetaData.msg is wholly adequate.

Reply:
I'm not sure I follow what you are saying. If we are using the raw data, then the costmap2d object will be constructed with the `costmap_raw_msg.metadata` (the occupancy grid msg isn't available in this case), where on the other hand it is the opposite if we are not using the raw topic.  So we can't use the msg that is not available.

Reply:
I'm saying these 2 blocks:

```
    if (costmap_ == nullptr) {
      costmap_ = std::make_shared<Costmap2D>(
        costmap_raw_msg_->metadata.size_x, costmap_raw_msg_->metadata.size_y,
        costmap_raw_msg_->metadata.resolution, costmap_raw_msg_->metadata.origin.position.x,
        costmap_raw_msg_->metadata.origin.position.y);
```

```
      if (costmap_ == nullptr) {
        costmap_ = std::make_shared<Costmap2D>(
          costmap_msg_->info.width, costmap_msg_->info.height,
          costmap_msg_->info.resolution, costmap_msg_->info.origin.position.x,
          costmap_msg_->info.origin.position.y);
```

Are exactly the same, if you have an if/else for grabbing a reference to the `MapMetaData`, ei `MapMetaData & my_meta` and then make the `costmap_` from `my_meta` then you reduce the duplication. But to do that the `Costmap.msg` needs to use the `nav_msgs/MapMetaData` rather than `nav2_msgs/CostmapMetaData` which are identical and we should completely remove the `nav2_msgs` redundancy. 

With that said, this isn't directly related to what you're trying to accomplish with this PR so its up to you if you want to do it or not. It would just make this logic cleaner and less repetitive.